### ISSUE TITLE
stb_image_write.h v1.15 - memory leak

### ISSUE THREAD

**Issue by Worst45** [nothings/stb#1126]
Hello,

using `stbi_write_jpg()` to generate a lot of JPG files in a row, I found a memory leak.

After digging in the code, it looks like in `stbi__end_write_file()`, `stbi__write_context.func` whas not released.
I fixed on my own by adding: `STBIW_FREE(s->func);`, like this:
```c
static void stbi__end_write_file(stbi__write_context *s)
{
   fclose((FILE *)s->context);
   STBIW_FREE(s->func);            // <-- added this line
}
```

Best regards

**Comment by nothings** [comment]
s->func is not allocated and should not be freed. It is not even data, it is a function pointer, and freeing it should crash.

the jpeg writer does not ever allocate memory, as far as I can see. check to make sure it's not your own code that's leaking.

**Comment by Worst45** [comment]
You must be right, as my C skills are not really good :)
But it seems to me the leak isn't something I can fix on my part of the code:

I only did:
```c
stbi_write_jpg(jpgAdr, IMAGE_WIDTH, IMAGE_HEIGHT, 1, image, quality);
```
 - This calls `stbi__start_write_file(&s,filename)`
 - which calls  `stbi__start_write_callbacks(s, stbi__stdio_write, (void *) f);`
 - which populate `->func` with a pointer to function `stbi__stdio_write`

And at the end of the call, I have more memory consumed. When I added the `STBIW_FREE()`, it returned the exact amount of RAM than before the call.

For me, the root cause was because of `stbi__write_context s = { 0 };` being destroyed at the end of the `stbi_write_jpg()` call, but still having a pointer to something in RAM.
So I thought it would be the right way to do, to clear the remaining pointer.

**Comment by nothings** [comment]
as far as I can see, only the PNG writer and the HDR writer call malloc. you can #define STBIW_MALLOC to your own function and verify that it's never called.

**Comment by nothings** [comment]
note that under the hood, it calls fopen / fclose, and fopen allocates memory, but fclose should free it.

**Comment by Worst45** [comment]
Found! It WAS on my side, finally :)

But it is very (very) strange, and I double-checked before answering here: it came from... my C compiler oO
I develop with Eclipse (on Windows), and use the "internal compiler". With it, I have the memory leak.

Just to be sure, I switched to MinGW. And then... no memory leak!
So, sorry for this false issue, it is related to the compiler, and not to the code...

and thanks a lot for your reactivity!
